daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/processors/DataProcessor.scala

```diff
@@ -340,8 +340,10 @@ class DataProcessor(
    * runtime. Instead we deal with success and failure statuses.
    */
   def parse(input: InputSourceDataInputStream, output: InfosetOutputter): DFDL.ParseResult = {
-    Assert.usage(!this.isError)
-
+    Assert.argCheck(
+      !this.isError,
+      s"Must check isError before calling DataProcessor parse. ${this.getDiagnostics.mkString}",
```

Comment:
This error message is making an assumption that the user isn't calling isError at all. But a user could call isError and have a bug where they just don't check the result correctly or call parse anyways.. So they might be confused--"but I am calling isError". Maybe a minor tweak like ?

> DataProcessor parse() must not be called if isError() returns true

It also feels odd to include diagnostics the message. Seems like this could hide the real issue of bad API usage. This could be more of an issue if there are alot of warnings mixed in, which might make it very hard to see this one like about bad API usage. Maybe we could filter the diags and show only the error? Might be even better to make the one error a `cuase` to the IllegalArgumentException. But I wonder if we should even do that? A user might see that diagnostic and fix the schema, when really they need to fix their code.

Reply:
The diagnostics were included based on this comment: https://github.com/apache/daffodil/pull/1058#pullrequestreview-1541309384, should I disregard?

Reply:
I don't feel strongly about whether we include Diagnostics or not. I'm still not convinced it makes sense, but if @mbeckerle or others think it's important then it's fine with me.

But I would suggest that it should be a `cause` to the `IllegalStateException` and not part of the exception message. The diagnostics are only a secondary reason for throwing this exception and don't belong in the message itself.

Maybe the `illegalState` function/macro accepts a third cause parameter, and it's called something like this:
```scala
Assert.illegalState(isError, "...", diagnostics.find(_.isError).get)
```

The macro would expand to
```scala
if (isError) {
  throw new IllegalStateException("...", diagnostics.find(_.isError).get)
}
```
So the message and `diagnostics.find` (or whatever logic is needed) would only be evaluated if `isError` is true.